rewrite-java/src/main/java/org/openrewrite/java/cleanup/ChainStringBuilderAppendCalls.java

```diff
@@ -0,0 +1,126 @@
+ * limitations under the License.
+ */
+package org.openrewrite.java.cleanup;
+
+import org.openrewrite.ExecutionContext;
+import org.openrewrite.Recipe;
+import org.openrewrite.TreeVisitor;
+import org.openrewrite.internal.lang.Nullable;
+import org.openrewrite.java.JavaIsoVisitor;
+import org.openrewrite.java.MethodMatcher;
+import org.openrewrite.java.search.UsesMethod;
+import org.openrewrite.java.tree.*;
+import org.openrewrite.java.utils.ExpressionUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static java.util.Collections.singletonList;
+
+public class ChainStringBuilderAppendCalls extends Recipe {
+    private static final MethodMatcher STRING_BUILDER_APPEND = new MethodMatcher("java.lang.StringBuilder append(String)");
+
+    @Override
+    public String getDisplayName() {
+        return "Chain `StringBuilder.append()` calls";
+    }
+
+    @Override
+    public String getDescription() {
+        return "String concatenation within calls to `StringBuilder.append()` causes unnecessary memory allocation. Except for concatenations of String literals, which are joined together at compile time. Replaces inefficient concatenations with chained calls to `StringBuilder.append()`.";
+    }
+
+    @Override
+    protected @Nullable TreeVisitor<?, ExecutionContext> getSingleSourceApplicableTest() {
+        return new UsesMethod<>(STRING_BUILDER_APPEND);
+    }
+
+    @Override
+    protected TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new JavaIsoVisitor<ExecutionContext>() {
+            @Override
+            public J.MethodInvocation visitMethodInvocation(J.MethodInvocation method, ExecutionContext ctx) {
+                if (STRING_BUILDER_APPEND.matches(method)) {
+                    List<Expression> arguments = method.getArguments();
+                    if (arguments.size() != 1) {
+                        return method;
+                    }
+
+                    List<Expression> flattenExpressions = new ArrayList<>();
+                    boolean flattenable = flatAdditiveExpressions(arguments.get(0), flattenExpressions);
+                    if (!flattenable) {
+                        return method;
+                    }
+
+                    if (flattenExpressions.stream().allMatch(exp -> exp instanceof J.Literal)) {
+                        return method;
+                    }
+
+                    // group expressions
+                    List<Expression> groups = new ArrayList<>();
+                    List<Expression> group = new ArrayList<>();
+                    for (Expression exp : flattenExpressions) {
+                        if (exp instanceof J.Literal) {
+                            group.add(exp);
+                        } else {
+                            addToGroups(group, groups);
+                            groups.add(exp);
+                        }
+                    }
+                    addToGroups(group, groups);
```

Comment:
The grouping also has some tricky edge cases. Rewriting `sb.append(op + 'a' + 'b')` as `sb.append(op).append('a' + `'b')` would not be correct.